### ISSUE TITLE
RFC: lake build: add --system-packages flag to use system package manager

### ISSUE THREAD

**Issue by a2379** [leanprover/lean4#5122]
### Proposal

In order for lean libraries to be packaged for Linux distributions, they must be installable using the system package manager. The system package manager will use `lake build` for building individual projects, but will manage the dependencies without lake. There should be a way for `lake build` to either:
1. build the project without considering dependencies (and assume they will be made available at runtime using `$LEAN_PATH` or some other method)
2. specify one or more package directories on the command-line for `lake` to use at build-time

Notes:
- The exact implementation (flag vs. environment variable, how to specify multiple paths, etc.) is not important, system package managers are flexible there.
- Patching lakefile.lean/toml manually for each library is not feasible.
- Fetching any information at build time from a remote endpoint (like the reservoir registry) or via git fetch is also not OK for reproducible package managers like Nix or Guix.
- Ideally lake would be able to link directly against compiled `.olean`/`.lean` files, and not require the original code or `lakefile.lean` for dependencies. Not a big deal if this isn't possible.

### Community Feedback

The system package manager problem was mentioned here, although the issue itself was for a different idea:
https://github.com/leanprover/lean4/issues/3193#issuecomment-1898009228

### Impact

The impact would be availability of Lean libraries on various Linux distributions.

I have Lean 4.10, batteries, and some other libraries packaged for Guix via a new `lean-build-system`. But I cannot package `mathlib` because lake keeps trying to `git fetch` the dependencies.

**Comment by Coda-Coda** [comment]
This sounds like a good idea to me! :smiley:

**Comment by divyaranjan1905** [comment]
Hello, has this been fixed?

**Comment by hargoniX** [comment]
No, it is also currently not recommended to package Lean with a system package manager as libraries currently rely on very specific releases and are often not forward or backward compatible. For example if you install Lean version 4.x and then work on a mathlib checkout that relies on 4.y you are just going to get lots of weird errors.

**Comment by divyaranjan1905** [comment]
> No, it is also currently not recommended to package Lean with a system package manager as libraries currently rely on very specific releases and are often not forward or backward compatible. For example if you install Lean version 4.x and then work on a mathlib checkout that relies on 4.y you are just going to get lots of weird errors.

I see, that makes sense, but in a transactional package management/distribution setting like Guix or Nix, the curl method doesn't work because it demands write access to read-only directories such as `~/.profile`.

**Comment by divyaranjan1905** [comment]
@hargoniX and other maintainers, at least you might guide us on how this might be implemented in a PR?  Because without this, mathlib et.al simply would be unusable in distros such as Guix.

**Comment by nomeata** [comment]
I believe there is on the roadmap the ability to vendor the dependencies, and then build without network access. This would presumably already help a bunch for your use case? (Can't comment on timelines or precise form of the feature.)

**Comment by divyaranjan1905** [comment]
> I believe there is on the roadmap the ability to vendor the dependencies, and then build without network access.

Indeed, that would be all we need.

**Comment by attila-lendvai** [comment]
i'm trying to update the (currently broken) lean4 package in guix. was there any progress on support for vendoring dependencies? any docs i should look at?

`/doc/make/index.md` doesn't seem to say anything about this.

**Comment by Coda-Coda** [comment]
@attila-lendvai I'm not aware of any progress on this yet, but others might know more. Hope the updating works out.

**Comment by nadja-y** [comment]
We implemented this for nixpkgs using Lake's existing `package-overrides.json` — no upstream changes needed. The full mathlib tree is packaged: NixOS/nixpkgs#497946

The same `package-overrides.json` mechanism should work for Guix. Reviews welcome.
